src/nvim/getchar.c

```diff
@@ -2312,7 +2317,7 @@ static int vgetorpeek(const int advance)
     }           /* if (!character from stuffbuf) */
 
     /* if advance is FALSE don't loop on NULs */
-  } while (c < 0 || (advance && c == NUL));
+  } while (!exiting);
```

Comment:
`exiting` is derived state from `c` and `advance`. When you introduce derived state you create the possibility of inconsistency --  e.g. `c < 0`, but `exiting` wasn't set correctly.